File: src/android/ArisePlugin.java
package io.arise;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;

/**
 * Arise Plugin
 * @author Guillaume Charhon - Arise.io
 *
 */
public class ArisePlugin extends CordovaPlugin{

	CallbackContext callbackContext;

	@Override
	/**
	 * Called by each javascript plugin function
	 */
	public boolean execute(String action, JSONArray data, final CallbackContext callbackContext) throws JSONException {
		this.callbackContext = callbackContext;
		// Check if the action has a handler
		Boolean isValidAction = true;

		try {
			// Initialize the framework
			if ("initialize".equals(action)) {
				String authKey = data.getString(0);
				String appName = data.getString(1);

				Arise.initialize(cordova.getActivity().getApplicationContext(), authKey, appName);

				callbackContext.success();

				return true;
				
			} else if ("getVariationWithListener".equals(action)) {
				// Retrieve a variation
				String tagName = data.getString(0);
				String defaultValue = data.getString(1);

				this.cordova.setActivityResultCallback(this);
				// Get and setup the variation
				ABTest.getVariationWithListener(tagName, defaultValue, new VariationListener() {
					@Override
					public void onVariationAvailable(String value) {
						// callback with the variation
						callbackContext.success(value);
					}
				});
				
			} else if ("recordView".equals(action)){
				// Record a view
				String tagName = data.getString(0);
				// record the view
				ABTest.recordView(tagName);
				callbackContext.success();
			} 
			else if ("recordConversion".equals(action)){
				String tagName = data.getString(0);
				// record the conversion
				ABTest.recordConversion(tagName);
				callbackContext.success();
			} else {
				// No handler for the action
				isValidAction = false;
			}
		} catch (IllegalStateException e){
			callbackContext.error(e.getMessage());
		} catch (JSONException e){
			callbackContext.error(e.getMessage());
		}

		// Method not found
		return isValidAction;

	}


}

File: src/ios/ArisePlugin.m
//
//  ArisePlugin.m
//
//  Created by Weeraphat Kaewnang on 06/09/13.
//
//

#import "ArisePlugin.h"
#import <Cordova/CDV.h>
#import <Arise/Arise.h>

@implementation ArisePlugin

- (void)initialize:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    // Get the arguments
    NSString* authKey = [command.arguments objectAtIndex:0];
    NSString* appName = [command.arguments objectAtIndex:1];
    
    if (authKey != nil && appName != nil) {
        // Iniatialize the Arise framework
        [Arise initializeWithKey:authKey appName:appName];
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)getVariationWithListener:(CDVInvokedUrlCommand*)command
{
    __block CDVPluginResult* pluginResult = nil;
    NSString* tagName = [command.arguments objectAtIndex:0];
    NSString* defaultValue = [command.arguments objectAtIndex:1];
    
    if (tagName != nil) {
        // Retrieve a variation
        [ABTest getVariation:tagName defaultValue:defaultValue data:^(NSString *experimentValue) {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:experimentValue];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        }];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
    }
}

- (void)recordView:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    NSString* tagName = [command.arguments objectAtIndex:0];
    if (tagName != nil) {
        // Record the view
        [ABTest recordView:tagName];
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}

- (void)recordConversion:(CDVInvokedUrlCommand*)command
{
    CDVPluginResult* pluginResult = nil;
    NSString* tagName = [command.arguments objectAtIndex:0];
    if (tagName != nil) {
        // Record the conversion
        [ABTest recordConversion:tagName];
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR];
    }
    [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
}


@end

